### PR TITLE
Adding the ability to handle "complex" point mutations

### PR DESCRIPTION
**Pull request by emarinier**
Adds the ability for starAMR to handle "complex" point mutations. That is, mutations where several individual point mutations are simultaneously required to confer a phenotype.

The solution has been created for Enterococcus faecium, but the framework should allow other similar mutations to be added to the "complex_mutations.tsv" file and handled in the same way.

### CODE REVIEW

File: staramr/blast/results/pointfinder/BlastResultsParserPointfinderResistance.py [hunk 1]
```diff
@@ -75,3 +78,26 @@ def _get_result(self, hit, db_mutation):
                 ]
 
         return result
+    
+    def _get_result_rows(self, hit, database_name):
+        database_resistance_mutations = self._get_resistance_mutations(hit, database_name)
```
[the remark this answers is not in the document]
Comment (emarinier) [reply staramr/blast/results/pointfinder/BlastResultsParserPointfinderResistance.py]: Yup. Updated!

File: staramr/tests/integration/detection/test_AMRDetection.py [hunk 2]
```diff
@@ -710,6 +711,152 @@ def testPointfinderSalmonella_16S_rrSD_C1065T_Success(self):
         self.assertEqual(expected_records['16S_rrsD'].seq.upper(), records['16S_rrsD'].seq.upper(),
                          "records don't match")
 
+
+    def testPointfinderEnterococcusFaecium_pbp5_Success(self):
```
Comment (apetkau) [inline staramr/tests/integration/detection/test_AMRDetection.py:715]: Could you add a test case where you don't detect all the required mutations and so there is no insertion of a "complex" resistance phenotype row?

**Review by apetkau — CHANGES_REQUESTED** [review]
Thanks so much for this @emarinier . Sorry for the delay in reviewing. This looks great. I tested it on a few example genomes :)

I just have a few comments, listed below.

File: staramr/databases/resistance/pointfinder/complex/data/complex_mutations.tsv [hunk 3]
```diff
@@ -0,0 +1,3 @@
+positions	mandatory	phenotype
+pbp5 (A216S), pbp5 (A499T), pbp5 (A68T), pbp5 (D204G), pbp5 (E100Q), pbp5 (E525D), pbp5 (E629V), pbp5 (E85D), pbp5 (G66E), pbp5 (K144Q), pbp5 (L177I), pbp5 (M485A), pbp5 (N496K), pbp5 (P667S), pbp5 (R34Q), pbp5 (S27G), pbp5 (T172A), pbp5 (T324A), pbp5 (V24A), pbp5 (V586L)	pbp5 (M485A), pbp5 (E629V), pbp5 (P667S)	Ampicillin
```
[the remark this answers is not in the document]
Comment (emarinier) [reply staramr/databases/resistance/pointfinder/complex/data/complex_mutations.tsv]: Updated in [148eb47](https://github.com/phac-nml/staramr/pull/172/commits/148eb47dc1e8b75015c7f36ac6a3cf46310c09d1).

File: staramr/results/AMRDetectionSummary.py [hunk 4]
```diff
@@ -5,6 +5,8 @@
 import pandas as pd
 from pandas import DataFrame
 
+pd.set_option('display.max_columns', None)
```
[the remark this answers is not in the document]
Comment (emarinier) [reply staramr/results/AMRDetectionSummary.py]: Yes, good catch. Removed in [cd11bbf](https://github.com/phac-nml/staramr/pull/172/commits/cd11bbfb7f81934a04b437919691cce787f87bc8).

File: staramr/tests/integration/detection/test_AMRDetection.py (hunk 2 above)
Comment (emarinier) [reply staramr/tests/integration/detection/test_AMRDetection.py:715]: Added in [08a9f4a](https://github.com/phac-nml/staramr/pull/172/commits/08a9f4a12d8624d8091109725d8a6ceef22bf8de).

**Comment by emarinier** [comment]
Updated the changelog in [2c7307c](https://github.com/phac-nml/staramr/pull/172/commits/2c7307c8f7aa5bbba1ede80560666004e300d89e).

File: staramr/databases/resistance/pointfinder/complex/ComplexMutations.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            if set(mandatory).issubset(set(mutation_codes)):
+                intersection = list(set(positions).intersection(set(mutation_codes)))
+
+                # Sorting, because it makes output reproducable.
+                # .sort() is in-place.
+                intersection.sort()
+
+                # Similar operation as above for positions, except we need to sort them
+                # as integers, rather than strings:      
+                mutation_positions = [re.sub("[^0-9]", "", i) for i in intersection]
```
[the remark this answers is not in the document]
Comment (apetkau) [reply staramr/databases/resistance/pointfinder/complex/ComplexMutations.py:49]: Thanks. That looks good. It was just a suggestion and it's all working as-is.

**Comment by apetkau** [comment]
Can you merge in the latest `development` to fix merge conflicts?

**Comment by emarinier** [comment]
Merge conflict should be resolved now.

**Review by apetkau — CHANGES_REQUESTED** [review]
Thanks Eric for merging in the code.

When I run staramr on data, I'm now getting this error:

```
2023-06-08 12:54:32 ERROR: 'NoneType' object has no attribute 'extend'
Traceback (most recent call last):
  File "/home/apetkau/workspace/staramr/bin/staramr", line 68, in <module>
    args.run_command(args)
  File "/home/apetkau/workspace/staramr/staramr/subcommand/Search.py", line 502, in run
    unacceptable_num_contigs= args.unacceptable_num_contigs)
  File "/home/apetkau/workspace/staramr/staramr/subcommand/Search.py", line 299, in _generate_results
    report_all_blast, ignore_invalid_files, mlst_scheme)
  File "/home/apetkau/workspace/staramr/staramr/detection/AMRDetection.py", line 199, in run_amr_detection
    plength_threshold_pointfinder, report_all)
  File "/home/apetkau/workspace/staramr/staramr/detection/AMRDetectionResistance.py", line 62, in _create_pointfinder_dataframe
    return pointfinder_parser.parse_results()
  File "/home/apetkau/workspace/staramr/staramr/blast/results/BlastResultsParser.py", line 67, in parse_results
    self._handle_blast_hit(file, database_name, blast_out, results, hit_seq_records)
  File "/home/apetkau/workspace/staramr/staramr/blast/results/BlastResultsParser.py", line 109, in _handle_blast_hit
    blast_results = self._get_result_rows(hit, database_name)
  File "/home/apetkau/workspace/staramr/staramr/blast/results/pointfinder/BlastResultsParserPointfinderResistance.py", line 90, in _get_result_rows
    results.extend(matches)
AttributeError: 'NoneType' object has no attribute 'extend'
```

Could you look into this?

Tested in Python 3.7. I haven't tested on other Python versions.

**Comment by emarinier** [comment]
Good catch, as always!

I think the issue was the complex mutation logic always assumed there would be existing results to extend, but sometimes there's no results. I added some logic to catch this.

**Pull request merged**